### ISSUE TITLE
E/SumUp: (SourceFile:153)@main | onTransportError() CardReaderError{Name = TRANSPORT_ERROR mErrorCode=3}

### ISSUE THREAD

**Issue by jordanscott1992** [sumup/sumup-android-sdk#150]
`W/SumUp: (SourceFile:112)@BtSmart-Receiver | BluetoothAdapter not initialized`
`E/SumUp: (SourceFile:153)@main | onTransportError() CardReaderError{Name = TRANSPORT_ERROR mErrorCode=3}`

Anyone have constant issues with Bluetooth connecting to the reader ?

**Comment by okirmis** [comment]
Hey @jordanscott1992 we are experiencing exactly the same issues. Sometimes it takes roughly 2 minutes until the customer is requested to enter the PIN and we're getting the same error message. Sometimes it fails completely. If I read the blinking LEDs on the SumUp Air Terminal correctly, it somehow communicates with the tablet.

SumUp team, anyone?

**Comment by jordanscott1992** [comment]
I eventually ditched the sumup reader for the izettle reader and swapped the sdk out.. No more Bluetooth issues it just works thankfully. 

Good luck getting a response, this is a known issue that involves clearing the Bluetooth cache on the device and resetting the reader.. Which makes it useless for my integration.

**Comment by okirmis** [comment]
Oh great, I'm looking forward to explaining that to our users. But thanks for responding so quickly!

**Comment by vladaman** [comment]
Yes, we are seeing something similar at https://github.com/sumup/sumup-android-sdk/issues/161

**Comment by dp-singh** [comment]
We had a similar issue: 

### To replicate this issue (8 out of 10 times Android Version 9 and 10) .

1. Power on  Android Device`  and `card reader` and do not connect card reader to power.
2. Connect app to card reader and do a transaction or cancel the transaction.
3. Exit app and power off the tablet. Do nothing on the card reader; it will go into standby mode after 2 minutes.
4. After card reader screen is blank (stand-by mode), power on tablet and try and do a transaction.
5. Observe multiple times we will face the issue of card reader not responding. `onTransportError() CardReaderError{Name = TRANSPORT_ERROR mErrorCode=3}`


### Stacktrace from nRG Connect  app which manage software for [Bluetooth Low Energy](https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp&hl=en_GB&gl=US)

![image](https://user-images.githubusercontent.com/2740228/136108359-f867609a-73a4-469c-8926-79fd49524409.png)

## Fix Validation

I started doing the `refresh` from `nRG Connect  app` before connecting to the `sumup sdk from our app`. And sumup sdk started working as desired.

![image](https://user-images.githubusercontent.com/2740228/136109469-9c2e5cac-ee7d-4cce-b41c-69711fcf4a2b.png)


### Fix code in the app.

On every app startup added code to refresh the `BluetoothGatt cache` using the https://github.com/Polidea/RxAndroidBle.


```
public class ClearCacheCustomOperation implements RxBleCustomOperation<Boolean> {
    @NonNull
    @Override
    public Observable<Boolean> asObservable(BluetoothGatt bluetoothGatt, RxBleGattCallback rxBleGattCallback, Scheduler scheduler) {
        return Observable.create(emitter -> {
            try {
                Method localMethod = bluetoothGatt.getClass().getMethod("refresh", new Class[0]);
                if (localMethod != null) {
                    boolean bool = (Boolean) localMethod.invoke(bluetoothGatt, new Object[0]);
                    emitter.onNext(bool);
                    emitter.onComplete();
                }
            } catch (Exception localException) {
                localException.printStackTrace();
                emitter.onNext(false);
            }
        });
    }
}
```


```
  private var scanDisposable: Disposable? = null
    private fun refreshBleDevice() {
        if (rxBleClient.isScanRuntimePermissionGranted) {
            scanBleDevices()
                .timeout(1, TimeUnit.MINUTES)
                .flatMap {
                    it.bleDevice
                        .establishConnection(false)
                        .flatMap { conn -> conn.queue(ClearCacheCustomOperation())}
                        .take(1)
                }
                .subscribe({Timber.d("Bluetooth config done")}, { Timber.e(it) })
                .let { scanDisposable = it }
        } 
    }

    private fun scanBleDevices(): Observable<ScanResult> {
        val scanSettings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .setCallbackType(ScanSettings.CALLBACK_TYPE_ALL_MATCHES)
            .build()

        val scanFilter = ScanFilter.Builder()
           .build()

        return rxBleClient.scanBleDevices(scanSettings, scanFilter)
    }
```

Hope that helps.

**Comment by adamschoenemann** [comment]
I'm having this issue as well! Very unfortunate. Will try the bluetooth cache refresh workaround, but it seems quite brittle.

**Comment by HeyPouya** [comment]
We have included improvements in the [version 6.0.0](https://github.com/sumup/sumup-android-sdk/releases/tag/v6.0.0) of the SDK. Can you please let us know if you still face this issue in this version?

**Comment by leila-sumup** [comment]
Closing this one due to lack of movement. 
If the issue still persists in later SDK versions, please reopen the issue and include context on the mobile used, reader used, reader firmware version and steps to reproduce so that we can investigate!
